Allow LCOV_EXCL to filter function coverage.

wsnyder opened the pull request:
This proposes to allow LCOV_EXCL_*'s to also filter function coverage.

I have some functions like

     void foo() override { assert_as_impossible(0); }

and desire a way to exclude them from the function totals, just like I can exclude them from line and branch at present.

(And with that think I'm done for a LONG time ;)

RPGillespie6 commented on fastcov.py:
```diff
@@ -282,6 +282,14 @@ def exclMarkerWorker(fastcov_sources, chunk, exclude_branches_sw, include_branch
+                for f in fastcov_data["functions"].keys():
+                    l = fastcov_data["functions"][f]["start_line"]
+                    if l not in line_to_func:
+                        line_to_func[l] = {}
+                    line_to_func[l][f] = f
```
```suggestion
                        line_to_func[l] = set()
                    line_to_func[l].add(f)
```
I recommend using a `set()` here instead of `dict()` since the the value is a don't care

RPGillespie6 commented on fastcov.py:
```diff
@@ -295,6 +303,10 @@ def exclMarkerWorker(fastcov_sources, chunk, exclude_branches_sw, include_branch
                             del fastcov_data[key][i]
+                    if i in line_to_func:
+                        for key in line_to_func[i]:
+                            if fastcov_data["functions"][key]:
```
I believe this will throw `KeyError` if key doesn't exist.
```suggestion
                            if key in fastcov_data["functions"]:
```

RPGillespie6 commented on fastcov.py:
```diff
@@ -309,6 +321,12 @@ def exclMarkerWorker(fastcov_sources, chunk, exclude_branches_sw, include_branch
                             if start_line <= line_num <= end_line:
                                 del fastcov_data[key][line_num]
 
+                    for i in range(start_line, end_line):
```
`i` is already declared and used by the outer loop. Better to use `line_num`
```suggestion
                    for line_num in range(start_line, end_line):
```

RPGillespie6 commented on fastcov.py:
```diff
@@ -309,6 +321,12 @@ def exclMarkerWorker(fastcov_sources, chunk, exclude_branches_sw, include_branch
+                    for i in range(start_line, end_line):
+                        if i in line_to_func:
+                            for key in line_to_func[i]:
+                                if fastcov_data["functions"][key]:
```
```suggestion
                                if key in fastcov_data["functions"]:
```

wsnyder commented:
> Was this PR meant as "inspiration" like the previous one or are you wanting to merge? 

Wasn't sure myself, but your providing good feedback decided the issue ;)

Think this addresses all of your issues, and improved tests so new code completely covered.

wsnyder commented:
Ran it through codacy just to make sure I didn't miss anything and it complained about missing periods, so added fixing them up to the patch set. Assuming this is acceptable you might prefer to rebase merge instead of squashing.

RPGillespie6 commented:
I like that the two changes are in separate commits, I'm just going to rebase and merge. Thanks for contributing!
